src/main/java/uk/gov/ida/verifyserviceprovider/services/UnifyingKeySignatureTrustEngine.java

```diff
@@ -0,0 +1,109 @@
+package uk.gov.ida.verifyserviceprovider.services;
+
+import com.google.common.collect.Lists;
+import net.shibboleth.utilities.java.support.resolver.CriteriaSet;
+import net.shibboleth.utilities.java.support.resolver.ResolverException;
+import org.opensaml.security.credential.Credential;
+import org.opensaml.security.credential.CredentialResolver;
+import org.opensaml.xmlsec.keyinfo.KeyInfoCredentialResolver;
+import org.opensaml.xmlsec.signature.support.impl.ExplicitKeySignatureTrustEngine;
+import uk.gov.ida.saml.metadata.EidasMetadataResolverRepository;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Extends ExplicitKeySignatureTrustEngine to match requests against Verify Hub signatures (from the
+ * ExplicitKeySignatureTrustEngine for the hub), and eIDAS signatures (from the provided EidasMetadataResolverRepository).
+ */
+public class UnifyingKeySignatureTrustEngine extends ExplicitKeySignatureTrustEngine {
+
+    public UnifyingKeySignatureTrustEngine(ExplicitKeySignatureTrustEngine hubTrustEngine, EidasMetadataResolverRepository eidasMetadataResolverRepository) {
+        super(
+            new UnifyingCredentialResolver(hubTrustEngine, eidasMetadataResolverRepository),
+            new UnifyingKeyInfoCredentialResolver(hubTrustEngine, eidasMetadataResolverRepository));
+    }
+
+    private static class UnifyingKeyInfoCredentialResolver implements KeyInfoCredentialResolver {
+
+        private final ExplicitKeySignatureTrustEngine hubTrustEngine;
+        private final EidasMetadataResolverRepository eidasMetadataResolverRepository;
+
+        private UnifyingKeyInfoCredentialResolver(ExplicitKeySignatureTrustEngine hubTrustEngine, EidasMetadataResolverRepository eidasMetadataResolverRepository) {
+            this.hubTrustEngine = hubTrustEngine;
+            this.eidasMetadataResolverRepository = eidasMetadataResolverRepository;
+        }
+
+        @Nonnull
+        @Override
+        public Iterable<Credential> resolve(@Nullable CriteriaSet criteria) throws ResolverException {
+            List<Credential> credentialsFound = new LinkedList<>();
+            if (criteria == null) { return credentialsFound; }
+
+            for (String eidasResolverEntityId : eidasMetadataResolverRepository.getResolverEntityIds()) {
+                KeyInfoCredentialResolver eidasCredentialResolver = eidasMetadataResolverRepository.getSignatureTrustEngine(eidasResolverEntityId).get().getKeyInfoResolver();
```

Comment:
What if `getSignatureTrustEngine(eidasResolverEntityId).get()` returns null? (`getSignatureTrustEngine` returns an Optional)

Reply:
My understanding is that it cannot, because `eidasMetadataResolverRepository` has recently supplied us with `eidasResolverEntityId`. However, I can probably make this a little more defensive against that case. 👍